File: pkg/transport/nats/nats.go
// Package nats is the NATS Core driver for AgentFlowBus.
//
// It maps the transport contract onto github.com/nats-io/nats.go: Publish,
// Subscribe (with optional queue groups), Request/Reply, and ephemeral
// _INBOX-backed Inbox for streaming responses.
package nats

import (
	"context"
	"errors"
	"sync"

	natsgo "github.com/nats-io/nats.go"

	"github.com/ModulationAI/agentflowbus/pkg/transport"
)

// Options configure the NATS driver.
type Options struct {
	URL  string
	Name string
}

// Option mutates Options.
type Option func(*Options)

// URL sets the NATS server URL (default "nats://localhost:4222").
func URL(u string) Option { return func(o *Options) { o.URL = u } }

// Name sets the client connection name reported to the server.
func Name(n string) Option { return func(o *Options) { o.Name = n } }

// Driver is the NATS Core transport. The zero value is unusable; construct
// with New and call Connect before any Publish/Subscribe/Request.
type Driver struct {
	opts Options

	mu     sync.Mutex
	conn   *natsgo.Conn
	ctx    context.Context
	cancel context.CancelFunc
}

// New constructs a Driver. Options are applied in order; later options win.
func New(opts ...Option) *Driver {
	o := Options{URL: natsgo.DefaultURL}
	for _, f := range opts {
		f(&o)
	}
	return &Driver{opts: o}
}

// Connect establishes the underlying NATS connection. Idempotent: a second
// call while already connected is a no-op.
func (d *Driver) Connect(_ context.Context) error {
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.conn != nil {
		return nil
	}
	var natsOpts []natsgo.Option
	if d.opts.Name != "" {
		natsOpts = append(natsOpts, natsgo.Name(d.opts.Name))
	}
	nc, err := natsgo.Connect(d.opts.URL, natsOpts...)
	if err != nil {
		return err
	}
	d.conn = nc
	d.ctx, d.cancel = context.WithCancel(context.Background())
	return nil
}

// Close drains the connection so in-flight handlers complete, then releases
// the cancel context handed to subscription handlers. Idempotent.
func (d *Driver) Close() error {
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.conn == nil {
		return nil
	}
	if d.cancel != nil {
		d.cancel()
	}
	err := d.conn.Drain()
	d.conn = nil
	return err
}

// Capabilities advertises NATS Core features. Persistence is false; durable
// streams are reserved for the future JetStream driver (v0.3).
func (*Driver) Capabilities() transport.Capabilities {
	return transport.Capabilities{
		Streaming:   true,
		Persistence: false,
		QueueGroup:  true,
		Headers:     true,
	}
}

func (d *Driver) Publish(_ context.Context, msg *transport.RawMessage) error {
	if msg == nil {
		return errors.New("nats: nil message")
	}
	nc, err := d.requireConn()
	if err != nil {
		return err
	}
	return nc.PublishMsg(toNats(msg))
}

func (d *Driver) Subscribe(_ context.Context, subject, queue string, h transport.Handler) (transport.Subscription, error) {
	if h == nil {
		return nil, errors.New("nats: nil handler")
	}
	nc, err := d.requireConn()
	if err != nil {
		return nil, err
	}

	driverCtx := d.ctx
	dispatch := func(m *natsgo.Msg) {
		// MsgHandler has no error return: handler errors are owned by the bus
		// middleware (recover/logging) layered above the transport. We pass
		// d.ctx so handlers learn about driver shutdown via cancellation.
		_ = h(driverCtx, fromNats(m))
	}

	var sub *natsgo.Subscription
	if queue == "" {
		sub, err = nc.Subscribe(subject, dispatch)
	} else {
		sub, err = nc.QueueSubscribe(subject, queue, dispatch)
	}
	if err != nil {
		return nil, err
	}
	return &natsSubscription{sub: sub}, nil
}

func (d *Driver) Request(ctx context.Context, msg *transport.RawMessage) (*transport.RawMessage, error) {
	if msg == nil {
		return nil, errors.New("nats: nil message")
	}
	nc, err := d.requireConn()
	if err != nil {
		return nil, err
	}
	resp, err := nc.RequestMsgWithContext(ctx, toNats(msg))
	if err != nil {
		return nil, err
	}
	return fromNats(resp), nil
}

func (d *Driver) OpenInbox(_ context.Context) (transport.Inbox, error) {
	nc, err := d.requireConn()
	if err != nil {
		return nil, err
	}
	subject := natsgo.NewInbox()
	box := &natsInbox{
		subject: subject,
		ch:      make(chan *transport.RawMessage, 64),
		done:    make(chan struct{}),
	}
	sub, err := nc.Subscribe(subject, func(m *natsgo.Msg) {
		select {
		case box.ch <- fromNats(m):
		case <-box.done:
			// Inbox closed mid-flight; drop the message rather than panic.
		}
	})
	if err != nil {
		return nil, err
	}
	box.sub = sub
	return box, nil
}

func (d *Driver) requireConn() (*natsgo.Conn, error) {
	d.mu.Lock()
	nc := d.conn
	d.mu.Unlock()
	if nc == nil {
		return nil, errors.New("nats: not connected")
	}
	return nc, nil
}

func toNats(msg *transport.RawMessage) *natsgo.Msg {
	nm := &natsgo.Msg{
		Subject: msg.Subject,
		Data:    msg.Data,
		Reply:   msg.ReplyTo,
	}
	if len(msg.Headers) > 0 {
		nm.Header = natsgo.Header{}
		for k, v := range msg.Headers {
			nm.Header.Set(k, v)
		}
	}
	return nm
}

func fromNats(m *natsgo.Msg) *transport.RawMessage {
	rm := &transport.RawMessage{
		Subject: m.Subject,
		Data:    m.Data,
		ReplyTo: m.Reply,
	}
	if len(m.Header) > 0 {
		rm.Headers = make(map[string]string, len(m.Header))
		for k := range m.Header {
			rm.Headers[k] = m.Header.Get(k)
		}
	}
	return rm
}

type natsSubscription struct {
	sub *natsgo.Subscription
}

func (s *natsSubscription) Unsubscribe() error { return s.sub.Unsubscribe() }

type natsInbox struct {
	subject string
	ch      chan *transport.RawMessage
	sub     *natsgo.Subscription
	done    chan struct{}
	once    sync.Once
}

func (i *natsInbox) Subject() string { return i.subject }

func (i *natsInbox) Recv(ctx context.Context) (*transport.RawMessage, error) {
	select {
	case m := <-i.ch:
		return m, nil
	case <-i.done:
		return nil, errors.New("nats inbox: closed")
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

func (i *natsInbox) Close() error {
	var err error
	i.once.Do(func() {
		close(i.done)
		if i.sub != nil {
			err = i.sub.Unsubscribe()
		}
	})
	return err
}

// Compile-time interface check.
var _ transport.Transport = (*Driver)(nil)

File: sdk/python/src/openagentio/transport/nats.py
"""NATS Core driver for OpenAgentIO. Mirrors pkg/transport/nats/nats.go.

Wraps the async ``nats-py`` client. Exposes Publish, Subscribe (with optional
queue groups), Request/Reply, and an _INBOX-backed :py:class:`Inbox` for
streaming responses. JetStream is reserved for v0.3.
"""
from __future__ import annotations

import asyncio
from typing import Any

import nats
from nats.aio.client import Client as NATSClient
from nats.aio.msg import Msg

from openagentio.transport.base import (
    Capabilities,
    Inbox,
    RawMessage,
    Subscription,
    Transport,
    TransportHandler,
)

_CLOSED = object()


class NATSDriver:
    """NATS Core async driver."""

    def __init__(
        self,
        url: str = "nats://localhost:4222",
        *,
        name: str = "",
        connect_timeout: float = 2.0,
    ) -> None:
        self._url = url
        self._name = name
        self._connect_timeout = connect_timeout
        self._nc: NATSClient | None = None
        self._lock = asyncio.Lock()

    async def connect(self) -> None:
        """Idempotent: a second call while already connected is a no-op."""
        async with self._lock:
            if self._nc is not None:
                return
            kwargs: dict[str, Any] = {
                "servers": [self._url],
                "connect_timeout": self._connect_timeout,
            }
            if self._name:
                kwargs["name"] = self._name
            self._nc = await nats.connect(**kwargs)

    async def close(self) -> None:
        """Drain in-flight handlers, then release the connection. Idempotent."""
        async with self._lock:
            nc = self._nc
            self._nc = None
        if nc is not None:
            try:
                await nc.drain()
            except Exception:
                pass

    def capabilities(self) -> Capabilities:
        return Capabilities(
            streaming=True,
            persistence=False,
            queue_group=True,
            headers=True,
        )

    async def publish(self, msg: RawMessage) -> None:
        if msg is None:
            raise ValueError("nats: nil message")
        nc = self._require_conn()
        await nc.publish(
            subject=msg.subject,
            payload=msg.data,
            reply=msg.reply_to or "",
            headers=msg.headers,
        )

    async def subscribe(
        self,
        subject: str,
        queue: str,
        handler: TransportHandler,
    ) -> Subscription:
        if handler is None:
            raise ValueError("nats: nil handler")
        nc = self._require_conn()

        async def cb(m: Msg) -> None:
            await handler(_from_nats(m))

        sub = await nc.subscribe(subject=subject, queue=queue or "", cb=cb)
        return _NATSSubscription(sub)

    async def request(
        self,
        msg: RawMessage,
        timeout: float | None = None,
    ) -> RawMessage:
        if msg is None:
            raise ValueError("nats: nil message")
        nc = self._require_conn()
        resp = await nc.request(
            subject=msg.subject,
            payload=msg.data,
            timeout=timeout if timeout is not None else 2.0,
            headers=msg.headers,
        )
        return _from_nats(resp)

    async def open_inbox(self) -> Inbox:
        nc = self._require_conn()
        subject = nc.new_inbox()
        return await _NATSInbox.create(nc, subject)

    async def flush(self) -> None:
        """Block until the server has acknowledged all queued protocol writes.

        Useful before exiting a process to ensure the last publish actually hit
        the wire, and in tests to make subscribe-then-publish on the same
        connection race-free.
        """
        nc = self._require_conn()
        await nc.flush()

    def _require_conn(self) -> NATSClient:
        if self._nc is None:
            raise RuntimeError("nats: not connected")
        return self._nc


def _from_nats(m: Msg) -> RawMessage:
    headers: dict[str, str] | None = None
    raw_headers = getattr(m, "headers", None) or getattr(m, "header", None)
    if raw_headers:
        headers = {str(k): str(v) for k, v in raw_headers.items()}
    return RawMessage(
        subject=m.subject,
        data=m.data,
        headers=headers,
        reply_to=m.reply or "",
    )


class _NATSSubscription:
    def __init__(self, sub: Any) -> None:
        self._sub = sub
        self._closed = False

    async def unsubscribe(self) -> None:
        if self._closed:
            return
        self._closed = True
        try:
            await self._sub.unsubscribe()
        except Exception:
            pass


class _NATSInbox:
    def __init__(self, subject: str) -> None:
        self._subject = subject
        self._queue: asyncio.Queue = asyncio.Queue()
        self._sub: Any | None = None
        self._closed = False

    @classmethod
    async def create(cls, nc: NATSClient, subject: str) -> "_NATSInbox":
        box = cls(subject)

        async def cb(m: Msg) -> None:
            if not box._closed:
                await box._queue.put(_from_nats(m))

        box._sub = await nc.subscribe(subject=subject, cb=cb)
        return box

    @property
    def subject(self) -> str:
        return self._subject

    async def recv(self, timeout: float | None = None) -> RawMessage:
        getter = self._queue.get()
        if timeout is not None:
            getter = asyncio.wait_for(getter, timeout)
        item = await getter
        if item is _CLOSED:
            await self._queue.put(_CLOSED)
            raise RuntimeError("nats inbox: closed")
        return item

    async def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        if self._sub is not None:
            try:
                await self._sub.unsubscribe()
            except Exception:
                pass
        await self._queue.put(_CLOSED)
